git/odb/tree.go

```diff
@@ -91,6 +92,63 @@ func (t *Tree) Encode(to io.Writer) (n int, err error) {
 	return
 }
 
+// Merge performs a merge operation against the given set of `*TreeEntry`'s by
+// either replacing existing tree entries of the same name, or appending new
+// entries in sub-tree order.
+//
+// It returns a copy of the tree, and performs the merge in O(n*log(n)) time.
+func (t *Tree) Merge(others ...*TreeEntry) *Tree {
+	unseen := make(map[string]*TreeEntry)
+
+	// Build a cache of name+filemode to *TreeEntry.
+	for _, other := range others {
+		key := fmt.Sprintf("%s%o", other.Name, other.Filemode)
```

Comment:
Ooh! What a tricky catch. I added a commit that should prevent this from happening in: 55d6a41.